File: demo/mem_res_fil.pl
#!/usr/bin/perl-w;
use strict;
use Getopt::Long;
my $lim=5;
my $qua=5;
my $splice=0.1;
GetOptions(
	'lim|l=i' => \$lim,
    'qua|q=i' => \$qua,
    'splice|s=s' => \$splice,
);
my($infile, $outfile, $logfile) = @ARGV;
my($tot, %cou, %inf, %res);
open (IN,$infile)||die"$!\n";
while (<IN>) {
	next if /^@/;
	chomp;
	my @spl = split(/\t/, $_);
	if ($spl[5] =~ /\*/) {
		$inf{$spl[0]} = 'Unm';
		next;
	}
	if ($spl[1] > 2000) {
		$res{Mul} += 1;
		next;
    }
	my($nhs, $len, $nmi) = (0, 0, 0);
	while ($spl[5] =~ /(\d+)[HS]/g) {
		$nhs += $1;
	}
	while ($spl[5] =~ /(\d+)[MISH]/g) {
		$len += $1;
       # print "####$len\n";
	}
	if ($spl[12] =~ /MD:.:(.+)/) {
		my $mid = $1;
		while ($spl[5] =~ /(\d+)D/g) {
			$mid =~ s/\^[ATCG]{$1}//;
		}
		$mid =~ s/\d+//g;
		$nmi = length($mid);
	}elsif ($spl[13] =~ /MD:.:(.+)/) {
		my $mid = $1;
		while ($spl[5] =~ /(\d+)D/g) {
			$mid =~ s/\^[ATCG]{$1}//;
		}
		$mid =~ s/\d+//g;
		$nmi = length($mid);
	}
	if (($nhs/$len) > $splice) {
		$inf{$spl[0]} = 'Spl';
	}elsif ($spl[4] < $qua) {
		$inf{$spl[0]} = 'Low';
	}elsif ($nmi > $lim) {
		$inf{$spl[0]} = 'Too';
	}
}
open (IN,$infile)||die"$!\n";
open (OUT, '>', $outfile)||die"$!\n";
open (LOG, '>', $logfile)||die"$!\n";
while (<IN>) {
	if (/^@/) {
		print OUT "$_";
		next;
	}
	chomp;
	my $out = $_;
	my @spl = split(/\t/, $_);
	next if $spl[1] > 2000;
	$tot += 1;
	if ($inf{$spl[0]} eq 'Unm') {
		$res{Unm} += 1;
	}elsif ($inf{$spl[0]} eq 'Spl') {
		$res{Spl} += 1;
	}elsif ($inf{$spl[0]} eq 'Low') {
		$res{Low} += 1;
	}elsif ($inf{$spl[0]} eq 'Too') {
		$res{Too} += 1;
	}else {
		$res{Pas} += 1;
		print OUT "$out\n";
	}
}
$res{Unm} += 0;
$res{Spl} += 0;
$res{Low} += 0;
$res{Too} += 0;
$res{Pas} += 0;
printf LOG ("Total_reads\t$tot\t100%\nPass_filter\t$res{Pas}\t%.2f%%\nUnmapped\t$res{Unm}\t%.2f%%\nMultiple_mapped\t$res{Mul}\t%.2f%%\nSplice\t$res{Spl}\t%.2f%%\nLow_map_quality\t$res{Low}\t%.2f%%\nToo_much_mismatch\t$res{Too}\t%.2f%%\n",$res{Pas}/$tot*100,$res{Unm}/$tot*100,$res{Mul}/$tot*100,$res{Spl}/$tot*100,$res{Low}/$tot*100,$res{Too}/$tot*100);
close LOG;
close OUT;

File: mem_res_fil.py
# coding=utf8
import re
import sys
import getopt


class Handler:
    def __init__(self, inputFile, outputFile, logFile, qua, lim, splice):
        self.inputFile = inputFile
        self.outputFile = outputFile
        self.logFile = logFile
        self.qua = qua
        self.lim = lim
        self.read = 0
        self.pass_filter = 0
        self.splice = splice
        self.result = {
            'Unm': 0,
            'Mul': 0,
            'Spl': 0,
            'Low': 0,
            'Too': 0
        }

    def run(self):
        with open(self.outputFile, 'w') as outputDatas:
            with open(self.inputFile, 'r') as inputDatas:
                line = inputDatas.readline()
                while line:
                    self.read += 1
                    if (self.analysis(line.replace('\n', ''))):
                        outputDatas.write(line)
                        self.pass_filter += 1
                    line = inputDatas.readline()

        self.makeLog()

    # True表示未被pass filter，False表示没有
    def analysis(self, line):
        datas = line.split('\t')
        nhs = 0
        len = 0
        nmi = 0
        if datas[5] == '*':
            self.result['Unm'] += 1
            return False
        if int(datas[1]) > 2000:
            self.result['Mul'] += 1
            return False
        for i in re.findall(r'(\d+)[HS]', datas[5]):
            nhs += int(i)
        for i in re.findall(r'(\d+)[MISH]', datas[5]):
            len += int(i)
        temp = re.match(r'MD:.:(.+)', datas[12]) or re.match(r'MD:.:(.+)', datas[13])
        if temp:
            mid = temp.group()
            for i in re.findall(r'(\d+)D', datas[5]):
                re.sub(r'[ATCG]{%s}' % i, '', mid)
            re.sub(r'\d+', '', mid)
            nmi = len(mid)

        if nhs / len > self.splice:
            self.result['Spl'] += 1
        elif int(datas[4]) < self.qua:
            self.result['Low'] += 1
        elif nmi > self.lim:
            self.result['Too'] += 1
        else:
            return True
        return False

    def makeLog(self):
        with open(self.logFile, 'w') as outputDatas:
            outputDatas.write(
                "Total_reads\t%s\t100%%\nPass_filter\t%s\t%s%%\nUnmapped\t%s\t%s%%\nMultiple_mapped\t%s\t%s%%\nSplice\t%s\t%s%%\nLow_map_quality\t%s\t%s%%\nToo_much_mismatch\t%s\t%s%%\n" % (
                self.read, self.pass_filter, self.pass_filter / self.read * 100, self.result['Unm'],
                self.result['Unm'] / self.read * 100, self.result['Mul'], self.result['Mul'] / self.read * 100, self.result['Spl'],
                self.result['Spl'] / self.read * 100, self.result['Low'], self.result['Low'] / self.read * 100, self.result['Too'],
                self.result['Too'] / self.read * 100))


def main():
    inputFile = ''
    outputFile = ''
    logFile = ''
    qua = 5
    lim = 5
    splice = 0.1
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hi:o:l:Q:L:S")
        for opt, arg in opts:
            if opt == '-h':
                print('input param: -i <inputfile> -o <outputfile> -l <logfile>')
                sys.exit()
            elif opt == '-i':
                inputFile = arg
            elif opt == '-o':
                outputFile = arg
            elif opt == '-l':
                logFile = arg
            elif opt == '-Q':
                qua = int(arg)
            elif opt == '-L':
                lim = int(arg)
            elif opt == '-S':
                splice = float(arg)
    except Exception as e:
        print(e)
        sys.exit(2)
    Handler(inputFile, outputFile, logFile, qua, lim, splice).run()

if __name__ == '__main__':
    main()